client/Wayland/wlf_cliprdr.c

```diff
@@ -0,0 +1,863 @@
+ */
+static UINT wlf_cliprdr_send_client_format_list(wfClipboard* clipboard)
+{
+	CLIPRDR_FORMAT_LIST formatList = { 0 };
+
+	formatList.msgFlags = CB_RESPONSE_OK;
+	formatList.numFormats = (UINT32)clipboard->numClientFormats;
+	formatList.formats = clipboard->clientFormats;
+	return clipboard->context->ClientFormatList(clipboard->context, &formatList);
+}
+
+static void wfl_cliprdr_add_client_format_id(wfClipboard* clipboard, UINT32 formatId)
+{
+	size_t x;
+	CLIPRDR_FORMAT* format;
+	const char* name = ClipboardGetFormatName(clipboard->system, formatId);
+
+	for (x=0; x<clipboard->numClientFormats; x++)
+	{
+		format = &clipboard->clientFormats[x];
+		if (format->formatId == formatId)
+			return;
+	}
+
+	format = realloc(clipboard->clientFormats, (clipboard->numClientFormats + 1) * sizeof(CLIPRDR_FORMAT));
+	if (!format)
+		return;
+
+	clipboard->clientFormats = format;
+	format = &clipboard->clientFormats[clipboard->numClientFormats++];
+	format->formatId = formatId;
+	format->formatName = NULL;
+	if (name && (formatId >= CF_MAX))
+		format->formatName = _strdup(name);
+}
+
+static void wlf_cliprdr_add_client_format(wfClipboard* clipboard, const char* mime)
+{
+	if (wlf_mime_is_html(mime))
+	{
+		UINT32 formatId = ClipboardGetFormatId(clipboard->system, "HTML Format");
+		wfl_cliprdr_add_client_format_id(clipboard, formatId);
+	}
+	else if (wlf_mime_is_text(mime))
+	{
+		wfl_cliprdr_add_client_format_id(clipboard, CF_TEXT);
+		wfl_cliprdr_add_client_format_id(clipboard, CF_OEMTEXT);
+		wfl_cliprdr_add_client_format_id(clipboard, CF_UNICODETEXT);
+	}
+	else if (wlf_mime_is_image(mime))
+	{
+		UINT32 formatId = ClipboardGetFormatId(clipboard->system, "image/bmp");
+		wfl_cliprdr_add_client_format_id(clipboard, formatId);
+		wfl_cliprdr_add_client_format_id(clipboard, CF_DIB);
+	}
+	wlf_cliprdr_send_client_format_list(clipboard);
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT wlf_cliprdr_send_data_request(wfClipboard* clipboard,
+        UINT32 formatId)
+{
+	CLIPRDR_FORMAT_DATA_REQUEST request;
+	ZeroMemory(&request, sizeof(CLIPRDR_FORMAT_DATA_REQUEST));
+	request.requestedFormatId = formatId;
+	return clipboard->context->ClientFormatDataRequest(clipboard->context,
+	        &request);
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT wlf_cliprdr_send_data_response(wfClipboard* clipboard, const BYTE* data,
+        size_t size)
+{
+	CLIPRDR_FORMAT_DATA_RESPONSE response;
+	ZeroMemory(&response, sizeof(CLIPRDR_FORMAT_DATA_RESPONSE));
+	response.msgFlags = (data) ? CB_RESPONSE_OK : CB_RESPONSE_FAIL;
+	response.dataLen = size;
+	response.requestedFormatData = data;
+	return clipboard->context->ClientFormatDataResponse(clipboard->context,
+	        &response);
+}
+
+BOOL wlf_cliprdr_handle_event(wfClipboard* clipboard, const UwacClipboardEvent* event)
+{
+	if (!clipboard || !event)
```

Comment:
Not sure these checks are needed.
Anyway the `if (!clipboard)` check is duplicated below.